src/connection.rs

```diff
@@ -0,0 +1,462 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
+// OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+use error::ConnectionError;
+use frame::{
+    codec::FrameCodec,
+    header::{ACK, ECODE_PROTO, FIN, Header, RST, SYN, Type},
+    Body,
+    Data,
+    Frame,
+    GoAway,
+    Ping,
+    RawFrame,
+    WindowUpdate
+};
+use futures::{prelude::*, self, stream::{Fuse, Stream as FuturesStream}, sync::{mpsc, oneshot}};
+use std::{collections::BTreeMap, sync::{atomic::AtomicUsize, Arc}, u32, usize};
+use stream::{self, Item, Stream, Window};
+use tokio_codec::Framed;
+use tokio_io::{AsyncRead, AsyncWrite};
+use Config;
+
+
+/// Connection mode
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
+pub enum Mode {
+    Client,
+    Server
+}
+
+
+// Commands sent from `Ctrl` to `Connection`.
+enum Cmd {
+    OpenStream(Option<Body>, oneshot::Sender<Stream>)
+}
+
+
+/// `Ctrl` allows controlling some connection aspects, e.g. opening new streams.
+#[derive(Clone)]
+pub struct Ctrl {
+    sender: mpsc::Sender<Cmd>
+}
+
+impl Ctrl {
+    fn new(sender: mpsc::Sender<Cmd>) -> Ctrl {
+        Ctrl { sender }
+    }
+
+    /// Open a new stream optionally sending some initial data to the remote endpoint.
+    pub fn open_stream(&self, data: Option<Body>) -> impl Future<Item=Stream, Error=ConnectionError> {
+        let (tx, rx) = oneshot::channel();
+        self.sender.clone()
+            .send(Cmd::OpenStream(data, tx))
+            .map_err(|_| ConnectionError::Closed)
+            .and_then(move |_| rx.map_err(|_| ConnectionError::Closed))
+    }
+}
+
+
+// Handle to stream. Used by connection to deliver incoming data.
+#[derive(Clone)]
+struct StreamHandle {
+    recv_win: Arc<Window>,
+    sender: mpsc::UnboundedSender<Item>,
+    ack: bool
+}
+
+
+struct Controller {
+    sender: Ctrl,
+    receiver: Fuse<mpsc::Receiver<Cmd>>
+}
+
+
+/// A connection which multiplexes streams to the remote endpoint.
+pub struct Connection<T> {
+    is_dead: bool,
+    mode: Mode,
+    resource: Framed<T, FrameCodec>,
+    config: Arc<Config>,
+    id_counter: usize,
+    streams: BTreeMap<stream::Id, StreamHandle>,
+    controller: Controller,
+    stream_rx: mpsc::UnboundedReceiver<(stream::Id, Item)>,
+    stream_tx: mpsc::UnboundedSender<(stream::Id, Item)>,
+    pending: Option<RawFrame>
+}
+
+impl<T> Connection<T>
+where
+    T: AsyncRead + AsyncWrite
+{
+    /// Create a new connection either in client or server mode.
+    pub fn new(resource: T, config: Arc<Config>, mode: Mode) -> Self {
+        debug!("new connection");
+        let controller = {
+            let (tx, rx) = mpsc::channel(1024);
```

Comment:
Isn't `unbounded` better?

Reply:
Why do you think so? I thought it would maybe be a good idea to exert some back-pressure if we are late opening a certain number of streams.